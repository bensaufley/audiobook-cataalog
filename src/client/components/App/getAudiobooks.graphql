query getAudiobooks($startId: ID) {
  getAudiobooks(start: $startId) {
    id
    name
    authors {
      id
      author {
        lastName
        firstName
      }
      meta
    }
  }
}
