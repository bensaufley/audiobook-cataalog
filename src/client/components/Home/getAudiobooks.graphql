query getAudiobooks($startId: ID) {
  getAudiobooks(start: $startId) {
    ...audiobook
  }
}

fragment audiobook on Audiobook {
  id
  duration
  filepath
  name
  year
  genres {
    name
  }
  authors {
    id
    author {
      lastName
      firstName
    }
    meta
  }
}
