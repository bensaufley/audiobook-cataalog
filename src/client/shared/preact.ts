import type { Ref } from 'preact';

// eslint-disable-next-line import/prefer-default-export
export const mergeRefs = <T extends HTMLElement>(...refs: (Ref<T> | null | undefined)[]): Ref<T> | null => {
  const filteredRefs = refs.filter((ref) => !!ref);
  if (!filteredRefs.length) return null;
  if (filteredRefs.length === 1) return filteredRefs[0]!;

  return (el: T | null) => {
    filteredRefs.forEach((ref) => {
      if (typeof ref === 'function') ref(el);
      // eslint-disable-next-line no-param-reassign
      else ref.current = el;
    });
  };
};
