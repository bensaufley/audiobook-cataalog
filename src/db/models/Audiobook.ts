import {
  type Association,
  type BelongsToManyAddAssociationMixin,
  BLOB,
  FLOAT,
  Model,
  type Optional,
  type Sequelize,
  STRING,
  UUID,
  UUIDV4,
} from 'sequelize';

import type models from '~db/models';
import type Author from '~db/models/Author';
import type { AuthorAttributes } from '~db/models/Author';
import type Narrator from '~db/models/Narrator';
import type { NarratorAttributes } from '~db/models/Narrator';
import type User from '~db/models/User';
import type UserAudiobook from '~db/models/UserAudiobook';
import type { UserAudiobookJSON } from '~db/models/UserAudiobook';

interface NullCoverProps {
  cover: null;
  coverType: null;
}

interface CoverProps {
  cover: Buffer;
  coverType: string;
}

export type AudiobookAttributes<HasCover extends boolean = boolean> = {
  id: string;
  title: string;
  filepath: string;
  duration: number | null;
} & (HasCover extends true ? CoverProps : HasCover extends false ? NullCoverProps : CoverProps | NullCoverProps);

type AudiobookCreationAttributes = Optional<AudiobookAttributes<boolean>, 'id'>;

export default class Audiobook<HasCover extends boolean = boolean> extends Model<
  AudiobookAttributes<HasCover>,
  AudiobookCreationAttributes
> {
  declare static associations: {
    Authors: Association<Audiobook, Author>;
    Narrators: Association<Audiobook, Narrator>;
    UserAudiobooks: Association<Audiobook, UserAudiobook>;
    Users: Association<Audiobook, User>;
  };

  public declare id: string;

  public declare title: string;

  public declare filepath: string;

  public declare cover: HasCover extends false ? null : HasCover extends true ? Buffer : Buffer | null;

  public declare coverType: HasCover extends false ? null : HasCover extends true ? string : string | null;

  public declare duration: number | null;

  public declare readonly createdAt: Date;

  public declare readonly updatedAt: Date;

  public declare Authors?: Author[];

  public declare Narrators?: Narrator[];

  public declare Users?: User[];

  public declare UserAudiobooks?: UserAudiobook[];

  public declare addAuthor: BelongsToManyAddAssociationMixin<Author, AuthorAttributes>;

  public declare addNarrator: BelongsToManyAddAssociationMixin<Narrator, NarratorAttributes>;

  public static associate(m: typeof models) {
    this.belongsToMany(m.Author, { through: m.AudiobookAuthor });
    this.belongsToMany(m.Narrator, { through: m.AudiobookNarrator });
    this.hasMany(m.UserAudiobook);
    this.belongsToMany(m.User, { through: m.UserAudiobook });
  }

  public static generate(sequelize: Sequelize) {
    return this.init(
      {
        id: {
          type: UUID,
          primaryKey: true,
          defaultValue: UUIDV4,
          allowNull: false,
          autoIncrement: false,
        },
        title: {
          type: STRING,
          allowNull: false,
        },
        filepath: {
          type: STRING,
          allowNull: false,
        },
        cover: {
          type: BLOB,
        },
        coverType: {
          type: STRING,
        },
        duration: {
          type: FLOAT,
        },
      },
      {
        modelName: 'Audiobook',
        sequelize,
      },
    );
  }
}

export interface AudiobookJSON<HasCover extends boolean = boolean>
  extends Omit<Audiobook<HasCover>, 'createdAt' | 'updatedAt' | 'UserAudiobooks'> {
  createdAt: string;
  updatedAt: string;
  UserAudiobooks: UserAudiobookJSON[];
}
