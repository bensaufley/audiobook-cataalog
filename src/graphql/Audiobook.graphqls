type Audiobook @entity(
  additionalFields: [
    { path: "meta.checksum", type: "string" }
  ]
) {
  id: ID! @id
  name: String! @column
  authors: [Author!]! @link
}
