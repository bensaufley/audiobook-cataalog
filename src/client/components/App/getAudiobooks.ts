// Generated file. Do not edit!

import * as Urql from '@urql/preact';
import gql from 'graphql-tag';

import type * as SchemaTypes from '~graphql/schema';

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
export type GetAudiobooksQueryVariables = SchemaTypes.Exact<{
  startId: SchemaTypes.Maybe<SchemaTypes.Scalars['ID']>;
}>;

export type GetAudiobooksQuery = {
  __typename?: 'Query';
  getAudiobooks: Array<{
    __typename?: 'Audiobook';
    id: string;
    name: string;
    authors: Array<{
      __typename?: 'AudiobookAuthor';
      id: string;
      meta: SchemaTypes.Maybe<string>;
      author: { __typename?: 'Author'; lastName: string; firstName: SchemaTypes.Maybe<string> };
    }>;
  }>;
};

export const GetAudiobooksDocument = gql`
  query getAudiobooks($startId: ID) {
    getAudiobooks(start: $startId) {
      id
      name
      authors {
        id
        author {
          lastName
          firstName
        }
        meta
      }
    }
  }
`;

export function useGetAudiobooksQuery(
  options: Omit<Urql.UseQueryArgs<GetAudiobooksQueryVariables>, 'query'> = {}
) {
  return Urql.useQuery<GetAudiobooksQuery>({ query: GetAudiobooksDocument, ...options });
}
