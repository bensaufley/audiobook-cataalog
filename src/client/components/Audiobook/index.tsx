import { Fragment, FunctionComponent, h } from 'preact';

import { AudiobookFragment } from '~client/components/Home/getAudiobooks';

const Audiobook: FunctionComponent<AudiobookFragment> = ({ name, cover, authors, year }) => (
  <div>
    {cover && <img src={cover} alt={name} />}
    <p>{name}</p>
    <p>
      by{' '}
      {authors.map(({ author: { firstName, lastName }, meta }, i) => (
        <>
          {i > 0 && ', '}
          {firstName} {lastName}
          {meta && ` (${meta})`}
        </>
      ))}
    </p>
    <p>{year}</p>
  </div>
);

export default Audiobook;
