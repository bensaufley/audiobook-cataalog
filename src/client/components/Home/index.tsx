import { FunctionComponent, h } from 'preact';

import { useGetAudiobooksQuery } from '~client/components/App/getAudiobooks';

const Home: FunctionComponent = () => {
  const [
    { data: { getAudiobooks: audioBooks } = { getAudiobooks: [] }, error },
  ] = useGetAudiobooksQuery();

  return (
    <div>
      <h1>Audiobook Catalog</h1>
      <pre>{JSON.stringify(error ? error.toString() : audioBooks, null, 2)}</pre>
    </div>
  );
};

export default Home;
