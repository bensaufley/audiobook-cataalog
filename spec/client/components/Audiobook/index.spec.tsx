/* eslint-disable react/jsx-props-no-spreading */
import { render } from '@testing-library/preact';
import { ObjectIDMock } from 'graphql-scalars';
import { h } from 'preact';

import Audiobook from '~client/components/Audiobook';
import { AudiobookFragment } from '~client/components/Home/getAudiobooks';
import theEchoWifeJSON from '~spec/__mocks__/data/[metadata] The Echo Wife.json';

describe('~client/components/Audiobook', () => {
  it('renders without exploding', () => {
    const audiobook: AudiobookFragment = {
      id: 'asdifpowaerew',
      duration: 32123141,
      filepath: 'the-power.m4b',
      name: 'The Power',
      year: 2017,
      genres: [],
      authors: [
        {
          id: 'aweirupawoeruewr',
          meta: null,
          author: { firstName: 'Naomi', lastName: 'Alderman' },
        },
      ],
    };

    const { container } = render(<Audiobook {...audiobook} />);

    expect(container).toMatchSnapshot();
  });

  it('renders real data without exploding', () => {
    const audiobook: AudiobookFragment = {
      id: ObjectIDMock(),
      duration: 2321432,
      filepath: 'the-echo-wife.m4b',
      name: theEchoWifeJSON.common.title,
      year: theEchoWifeJSON.common.year,
      genres: [{ name: 'fiction' }, { name: 'science fiction' }],
      authors: [
        {
          id: ObjectIDMock(),
          meta: null,
          author: { firstName: 'Naomi', lastName: 'Alderman' },
        },
      ],
    };

    const { container } = render(<Audiobook {...audiobook} />);

    expect(container).toMatchSnapshot();
  });
});
