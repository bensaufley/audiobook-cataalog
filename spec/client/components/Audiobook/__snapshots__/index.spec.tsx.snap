// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`~client/components/Audiobook renders real data without exploding 1`] = `
<div>
  <div
    style="background-image: url(/cover/5e5677d71bdc2ae76344968c);"
    tabindex="0"
  >
    <div>
      <h3>
        <strong>
          The Echo Wife
        </strong>
         (2021)
      </h3>
      <p>
        644:50:32
        <br />
        by
         
        Naomi
         
        Alderman
      </p>
      <a
        href="/downloads/the-echo-wife.m4b"
      >
        Download
      </a>
    </div>
  </div>
</div>
`;

exports[`~client/components/Audiobook renders without exploding 1`] = `
<div>
  <div
    style="background-image: url(/cover/asdifpowaerew);"
    tabindex="0"
  >
    <div>
      <h3>
        <strong>
          The Power
        </strong>
         (2017)
      </h3>
      <p>
        8923:05:41
        <br />
        by
         
        Naomi
         
        Alderman
      </p>
      <a
        href="/downloads/the-power.m4b"
      >
        Download
      </a>
    </div>
  </div>
</div>
`;
